### PR TITLE
Implement bug fixes, add new tests

### PR DESCRIPTION
**Pull request by tadgh**
* Swap from a search by PID (invalid in case of forced IDs), to sequential reads during write phase. This corrects a bug in exported results.
* Lower batch size to 900 to avoid oracle clause count.
* Fix bug in group member resolution which was causing incorrect expansions. 
* Add more use case tests
* Changelogs

### CODE REVIEW

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/ExpandResourcesStep.java [hunk 1]
```diff
@@ -98,24 +93,24 @@ public RunOutcome run(@Nonnull StepExecutionDetails<BulkExportJobParameters, Bul
...
 			idList.getResourceType());
 
 		// and return
 		return RunOutcome.SUCCESS;
 	}
 
-	private IBundleProvider fetchAllResources(BulkExportIdList theIds) {
+	private List<IBaseResource> fetchAllResources(BulkExportIdList theIds) {
 		IFhirResourceDao<?> dao = myDaoRegistry.getResourceDao(theIds.getResourceType());
+		List<IBaseResource> resources = new ArrayList<>();
```
Comment (michaelabuckley) [inline hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/ExpandResourcesStep.java:107]: Maybe add a comment 
```suggestion
  // This should be a query, but we have PIDs, and we don't have a _pid search param
```

**Review by michaelabuckley — APPROVED** [review]

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/bulk/export/model/ExportPIDIteratorParameters.java [hunk 2]
```diff
@@ -122,4 +122,16 @@ public void setExpandMdm(boolean theExpandMdm) {
 		myExpandMdm = theExpandMdm;
 	}
 
+	@Override
+	public String toString() {
+		return "ExportPIDIteratorParameters{" +
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/bulk/export/model/ExportPIDIteratorParameters.java:127]: nice, fixed.

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/ExpandResourcesStep.java (hunk 1 above)
Comment (tadgh) [reply hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/ExpandResourcesStep.java:107]: Fixed

File: hapi-fhir-jpaserver-test-utilities/src/test/resources/logback-test.xml [hunk 3]
```diff
@@ -63,6 +63,10 @@
 		<appender-ref ref="STDOUT" />
 	</logger>
 
+	<logger name="ca.uhn.fhir.jpa.bulk" additivity="false" level="trace">
+		<appender-ref ref="STDOUT" />
+	</logger>
+
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-jpaserver-test-utilities/src/test/resources/logback-test.xml:69]: Fixed

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/SearchParameterUtil.java [hunk 4]
```diff
@@ -98,9 +100,22 @@ public static RuntimeSearchParam getOnlyPatientCompartmentRuntimeSearchParam(Run
...
 	}
 
-	public static List<RuntimeSearchParam> getAllPatientCompartmentRuntimeSearchParams(FhirContext theFhirContext, String theResourceType) {
+	public static List<RuntimeSearchParam> getAllPatientCompartmentRuntimeSearchParamsForResourceType(FhirContext theFhirContext, String theResourceType) {
 		RuntimeResourceDefinition runtimeResourceDefinition = theFhirContext.getResourceDefinition(theResourceType);
 		return getAllPatientCompartmentRuntimeSearchParams(runtimeResourceDefinition);
+	}
+
+	public static List<RuntimeSearchParam> getAllPatientCompartmenRuntimeSearchParams(FhirContext theFhirContext) {
+		return theFhirContext.getResourceTypes()
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/util/SearchParameterUtil.java:109]: Yup, slow, but cached where needed.

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_1_2/4022-bulk-export-woes-4.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+---
+type: add
+issue: 4022
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_1_2/4022-bulk-export-woes-4.yaml:3]: image

My soul leaving my body every time i find a new bulk export bug

**Pull request merged**